risc0/zkvm/src/guest/env/mod.rs

```diff
@@ -500,6 +500,7 @@ pub fn read_buffered<T: DeserializeOwned>() -> Result<T, crate::serde::Error> {
 
 /// get an updated keccak state
 #[cfg(feature = "unstable")]
-pub fn keccak_update(state: &mut risc0_circuit_keccak::KeccakState) {
+#[no_mangle]
+pub fn risc0_keccak_update(state: &mut risc0_circuit_keccak::KeccakState) {
```

Comment:
Since misuse of this function could cause undefined behavior, it should be marked as unsafe. A function that uses `unsafe` internally should only be marked as safe (i.e. not marked `unsafe`) when it internally ensures that memory safety is maintained. (e.g. a deserializer may chose to use `unsafe` internally for performance, but makes sure to check buffer lengths and pointer to ensure safety for anyone using its public API).

If we are making an argument that this is safe because it can only be used inside the zkVM, and this is safe because the zkVM is single-threaded and non-preemptive, then we might want to note this.

```suggestion
pub unsafe fn risc0_keccak_update(state: &mut risc0_circuit_keccak::KeccakState) {
```

Reply:
I'm not sure we should be using `unsafe` in this manner. Is there any way a user could call this function and wind up with undefined behavior?

Reply:
If the zkVM makes provides as a contract that it will not allow preemptive or concurrent execution, and this function is only allowed in the zkVM context, then I think we can argue it is safe and mark it as such. I think that may be the case here, but we don't provide any explanation for why this function, at an API level is safe while using `unsafe` code internally.

Reply:
I think for every `unsafe` we should have a `* SAFETY` section.